### PR TITLE
*ClusterDescription → ReconciliationState

### PR DESCRIPTION
**Pull request by tombentley**
### Type of change

- Refactoring

### Description

This PR refactors the `*ClusterDescription` static member classes of `KafkaAssemblyOperator` into a single `ReonciliationState` member class. This will make it much simpler to pass cluster-wide state to and between the different methods participating in the reconciliation (necessary for things like CA renewal).

The change to `OpenSslCertManagerTest` is simply a fix to a test which happened to fail for me due (I assume) to some other process using `/tmp` during the test.

### Checklist

- [ ] Write tests
- [x] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

**Comment by tombentley** [comment]
* Having separate classes made it difficult to pass around cluster-wide state. Because of the way these things get used with `Future` it's not so easy to pass such state from one class to another (and doing that would be more complicated). For the cluster CA secrets this was worked around using a parameter (`List<Secret>`), but it was intrusive to have to pass that all over the place, and the situation would get worse when there's more *mutable* state to pass around (e.g. state concerning CA renewal).

  (It's worth pointing out that we could still factor this into separate methods if there is a need to categorise the different things we do (ZK, Kafka, EO). I wasn't convinced that was necessarily better than just having a single `compose()` chain, but I'm not convinced it's worse either.)

  It's also worth pointing out that the fields in `ReconciliationState` are a reflection of the 'awkward bits' of the process. What I mean by that is that I've removed fields which we previously just holding the result of a `*Cluster.generate*()` call and which was only going to be used in a single place, and instead just called that `*Cluster.generate*()` in that place. This means that all the remaining fields have some other non-trival logic associated with them, and therefore their presence *as fields* helps make that complexity clearer.

* The class was `static` before, but by making at a normal (instance) member class it captures the `KafkaAssemblyOperator`s fields (all the resource operators), which simplifies things a lot as I don't need parameters for those in each method called in the `compose()` chain. If I separated it into a separate class I would need fields for all of those and another huge constructor. It's not clear to me that that would bring any huge benefit either, because the class cannot be reused outside `KafkaAssemblyOperator`.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java [hunk 1]
```diff
@@ -94,372 +93,479 @@ public KafkaAssemblyOperator(Vertx vertx, boolean isOpenShift,
     @Override
     public Future<Void> createOrUpdate(Reconciliation reconciliation, Kafka kafkaAssembly, List<Secret> assemblySecrets) {
 
-        return createOrUpdateZk(reconciliation, kafkaAssembly, assemblySecrets)
-            .compose(i -> createOrUpdateKafka(reconciliation, kafkaAssembly, assemblySecrets))
-            .compose(i -> createOrUpdateTopicOperator(reconciliation, kafkaAssembly, assemblySecrets))
-            .compose(i -> createOrUpdateEntityOperator(reconciliation, kafkaAssembly, assemblySecrets));
+        String name = kafkaAssembly.getMetadata().getName();
+        log.debug("{}: create/update zookeeper {}", reconciliation, name);
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: We should fix the log message -> It is not updating ZK only now.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java [hunk 2]
```diff
@@ -94,372 +93,479 @@ public KafkaAssemblyOperator(Vertx vertx, boolean isOpenShift,
...
+        Future<Void> chainFuture = Future.future();
+        new ReconciliationState(kafkaAssembly).getZookeeperState(assemblySecrets)
+                .compose(desc -> desc.zkScaleDown())
```
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: Should `desc` be renamed in `state`? Wdyt?

**Comment by tombentley** [comment]
>  I have just a concern about readability of the createOrUpdate method which is now huge and we are using just blank lines to separate blocks related to zk, kafka and so on.

As already mentioned, we could factor this into 3 or 4 different methods, but it's not really clear to me that that would really be so much better. It would mean each method was smaller, which is a plus, but it would make the overall process less clear. I think I'd like to leave this for now, until I've done the next couple of changes which will result in propagating cluster CA state through the chain. We should then be in a better position to decide how best to improve it.

**Pull request merged**